http/peer-id-auth.md

```diff
@@ -0,0 +1,101 @@
+# Peer ID Authentication over HTTP
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2023-01-23  |
+
+Authors: [@MarcoPolo]
+
+Interest Group: Same as [HTTP](README.md)
+
+## Introduction
+
+This spec defines one way of authenticating Peer IDs over HTTP using a
+challenge-response scheme.
+
+## Mutual Client and Server Peer ID Authentication
+
+1. The server initiates the authentication by responding to a request that must
+   be authenticated with the response header `WWW-Authenticate: Libp2p-Challenge
+   challenge="<base64-encoded-challenge>, Libp2p-Challenge-Server-Only"`. The
```

Comment:
Why not just name them `-Client` and `-Server`? Aren't the fully composable?

Like, wouldn't it be possible to only authenticate the client for example?

```suggestion
   be authenticated with the response header `WWW-Authenticate: Libp2p-Challenge-Client
   challenge="<base64-encoded-challenge>, Libp2p-Challenge-Server"`. The
```

Reply:
These don't exactly compose in the way I think you are thinking about. It's not standard to include two credentials in an `Authorization` header (which I think is required for the composition you want).

`Libp2p-Challenge` authenticates both the client and server. It's different than a "authenticate client only mode". We could introduce a new mode to only authenticate the client (`Libp2p-Challenge-Client`), but what's the use case? When would a server not have a key to authenticate with, but the client does?

I would suggest not doing it now and adding it later, if a good use case arises.

Reply:
> > It's not standard to include two credentials in an `Authorization` header (which I think is required for the composition you want).
> 
> Unless I am misunderstanding the article, MDN says something different: [developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate#syntax](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate#syntax)
> 
> I think the following would be entirely legitimate (I renamed a few things to make them more concise):
> 
> ```
> WWW-Authenticate: Libp2p-Client challenge=<base64-encoded-challenge>
> WWW-Authenticate: Libp2p-Server challenge=<base64-encoded-challenge>
> ```

I just realized that this provides the client with two _options_ of how to authenticate instead of requiring them both. In that case, we probably need to go with the single scheme and two parameters:

```
> WWW-Authenticate: Libp2p server-challenge=<base64-encoded-challenge>, client-challenge=<base64-encoded-challenge>
```

Nit: I am unsure about the naming of the auth-scheme. It would be great if we can somehow reference _this_ particular spec with it. Perhaps it should be called `Libp2p-PeerId`? I find `Libp2p-Challenge` verbose and not very descriptive because `WWW-Authenticate` itself already specifies that its value is some kind of challenge.

Reply:
> > It's not standard to include two credentials in an `Authorization` header (which I think is required for the composition you want).
> 
> Unless I am misunderstanding the article, MDN says something different: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/WWW-Authenticate#syntax
> 
> I think the following would be entirely legitimate (I renamed a few things to make them more concise):
> 
> ```
> WWW-Authenticate: Libp2p-Client challenge=<base64-encoded-challenge>
> WWW-Authenticate: Libp2p-Server challenge=<base64-encoded-challenge>
> ```

The server can provide multiple authentication challenge kind (e.g. basic and bearer), but the client can only respond with a single scheme in `Authorization`. The [RFC section for `WWW-Authenticate`](https://www.rfc-editor.org/rfc/rfc9110#section-11.6.1) says, as you say, any number of challenges are allowed (`WWW-Authenticate = #challenge`). But for the [`Authorization`](https://www.rfc-editor.org/rfc/rfc9110#section-11.6.2), it's a single credential (`Authorization = credentials`).

But that's tangential. I think we both agree we should have a single auth scheme that has multiple parameters. So I'll change the spec to be that.